File: code/c++/GC.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <map>
#include <vector>
#include <iomanip>

class GCSolution
{
private:
    std::vector<std::string> labels;
    std::vector<double> gc_content_freq;

public:
    GCSolution()
    {
        read_dna_fasta_format();
    }

    // calcular a frequência GC no DNA
    double count_gc_content_freq(std::string dna)
    {
        double quantity = 0;

        for (auto base : dna)
        {
            if (base == 'G' || base == 'C')
            {
                quantity++;
            }
        }

        double freq = 100 * quantity / dna.length();

        return freq;
    }

    // lendo o DNA no formato FASTA
    void read_dna_fasta_format()
    {
        std::ifstream input_data("C:/Programming/Rosalind/input/rosalind_gc.txt");
        std::string line, dna_seq;
        double frequency;

        if (input_data.is_open())
        {
            // lendo a linha de texto
            while (std::getline(input_data, line, '\n'))
            {

                // verificando se a linha é o rótulo da sequência de DNA
                if (line[0] == '>')
                {
                    // adicionando os rótulo
                    labels.push_back(line.substr(1));

                    // adicionando a sequência de DNA
                    if (!dna_seq.empty())
                    {
                        frequency = count_gc_content_freq(dna_seq);
                        gc_content_freq.push_back(frequency);
                        dna_seq.clear();
                    }

                    continue;
                }

                dna_seq += line;
            }
            // adicionando a última sequência de DNA lida
            frequency = count_gc_content_freq(dna_seq);

            gc_content_freq.push_back(frequency);

            input_data.close();
        }
        else
        {
            std::cerr << "Não foi possível abrir o arquivo. " << std::endl;
        }
    }

    // verificar o índice da maior frequência
    unsigned check_max(std::vector<double> frequency)
    {
        double high = frequency[0];
        unsigned pos = 0;
        for (unsigned i = 1; i < frequency.size(); ++i)
        {
            if (frequency[i] > high)
            {
                high = frequency[i];
                pos = i;
            }
        }
        return pos;
    }

    void get_highest_gc_content()
    {
        unsigned i = check_max(gc_content_freq);

        std::ofstream output_data("C:/Programming/Rosalind/output/GC.txt");

        if (output_data.is_open())
        {
            output_data << labels[i] << '\n';
            output_data << std::fixed << std::setprecision(6) << gc_content_freq[i] << '\n';
        }
        else
        {
            std::cerr << "Não foi possível abrir o arquivo. " << std::endl;
        }
    }
};

int main()
{
    GCSolution solution;
    solution.get_highest_gc_content();

    return 0;
}

File: code/python/GC.py
import os


class GCSolution:
    def __init__(self):
        self.sequences_of_dna = []
        self.labels = []
        self.quantities = []

        self.read_sequences_dna()

    # lendo as sequências de dna do arquivo
    def read_sequences_dna(self):
        try:
            path = os.path.realpath("../../input/rosalind_gc.txt")
            input_data = open(path, 'r')
            # linhas de texto
            lines = input_data.readlines()
            # lista com todos os rótulos
            self.labels = [label for label in lines if label[0] == '>']

            dna_seq = ""
            for line in lines:
                # verificando se a linha de texto não é um rótulo
                if line not in self.labels:
                    dna_seq += line.strip()
                    # verificando se é a última linha
                    if line == lines[-1]:
                        # adicionando a última seuquência de dna
                        self.sequences_of_dna.append(dna_seq)
                    continue

                if dna_seq:
                    # adicioando as sequências de dna
                    self.sequences_of_dna.append(dna_seq)
                dna_seq = ""
            input_data.close()
        except IOError:
            print("Erro na leitura do arquivo!")

    # pegando a quantidade do conteúdo GC na sequência de DNA
    def get_gc_content(self, dna: str):
        quantity = 0
        for base in dna:
            if base in ['G', 'C']:
                quantity += 1
        return quantity

    # pegando o índice da maior quantidade
    def get_index_largest_quantity(self, quantities: list):
        max = quantities[0]
        index = 0
        for i in range(1, len(quantities)):
            if quantities[i] > max:
                max = quantities[i]
                index = i
        return index

    # pegando a maior quantidade
    def get_highest_gc_content(self):
        for dna in self.sequences_of_dna:
            quantity = self.get_gc_content(dna)
            self.quantities.append(quantity*100/len(dna))
        i = self.get_index_largest_quantity(self.quantities)
        return i

    # maior quantidade
    def highest_gc_content(self):
        i = self.get_highest_gc_content()
        quantity = self.quantities[i]
        label = self.labels[i]

        try:
            path = os.path.realpath("../../output/GC.txt")
            output_data = open(path, 'w')
            output_data.write(f"{label}")
            output_data.write(f"{quantity:.6f}")
            output_data.close()
        except IOError:
            print("Erro na escrita do arquivo!")


if __name__ == "__main__":
    solution = GCSolution()
    solution.highest_gc_content()
